extmod/network_lwip.c

```diff
@@ -90,6 +100,241 @@ mp_obj_t mod_network_nic_ifconfig(struct netif *netif, size_t n_args, const mp_o
+                                for (i = 1; i < LWIP_IPV6_NUM_ADDRESSES; i++) {
+                                    if (ip6_addr_isvalid(netif_ip6_addr_state(netif, i)) &&
+                                        !netif_ip6_addr_isstatic(netif, i)) {
+                                        found = 1;
+                                        break;
+                                    }
+                                }
+                                if (found) {
+                                    break;
+                                }
+                                if (mp_hal_ticks_ms() - start > 10000) {
+                                    mp_raise_msg(&mp_type_OSError, MP_ERROR_TEXT("timeout waiting for DHCP to get IP address"));
+                                }
+                                mp_hal_delay_ms(100);
+                            }
+                        } else {
+                            // Skip link-local address in slot 0
+                            for (i = 1; i < LWIP_IPV6_NUM_ADDRESSES; i++) {
+                                if (ip6_addr_isvalid(netif_ip6_addr_state(netif, i)) &&
+                                    !netif_ip6_addr_isstatic(netif, i)) {
+                                    netif_ip6_addr_set_state(netif, i, IP6_ADDR_INVALID);
+                                }
+                            }
+                        }
+                        break;
+                    }
+                    #endif
+                    case MP_QSTR_dns: {
+                        ip_addr_t dns;
+                        size_t addr_len;
+                        const char *addr_str = mp_obj_str_get_data(e->value, &addr_len);
+                        if (!ipaddr_aton(addr_str, &dns)) {
+                            mp_raise_ValueError(MP_ERROR_TEXT("invalid arguments as dns server"));
+                        }
+                        dns_setserver(0, &dns);
+                        break;
+                    }
+                    #if LWIP_VERSION_MAJOR >= 2
+                    case MP_QSTR_prefer: {
+                        int value = mp_obj_get_int(e->value);
+                        if (value != 4 && value != 6) {
+                            mp_raise_ValueError(MP_ERROR_TEXT("invalid prefer argument"));
+                        }
+                        mp_mod_network_prefered_dns_ip = value;
+                        break;
+                    }
+                    #endif
+                    case MP_QSTR_addr4:
+                    case MP_QSTR_addr6: {
+                        ip_addr_t ip_addr;
+                        int prefix = 32;
+                        if (e->value != mp_const_none) {
+                            size_t addr_len;
+                            const char *addr_str = mp_obj_str_get_data(e->value, &addr_len);
+                            char *split = strchr(addr_str, '/');
+                            if (split) {
+                                *split = 0;
```

Comment:
Unfortunately this is mutating the original string in `e->value`, as `mp_obj_str_get_data` does not copy the data. (And will outright fail if the string is in ROM, e.g. frozen str or qstr).